README.md
SUPReMM Processing Tools
========================

Our team at the Center for Computational Research University at Buffalo
develop and support a range of tools for the comprehensive management of
advanced cyber infrastructure (CI) resources, including high performance
computing (HPC). Part of comprehensive CI management is the monitoring and analysis
of user's HPC jobs. The suite of tools to support job-level performance
analysis was originally developed under a project called "SUPReMM".

The SUPReMM architecture comprises three major components:

* Software that runs directly on HPC compute nodes and periodically collects performance information.
* Software that uses the node-level performance data to generate job-level data.
* An [Open XDMoD][xdmod] module that enables the job-level information to be viewed and analyzed.

This repository contains the software that combines the node-level performance
data to generate job-level summary data.

Full details of the SUPReMM project are available on the [SUPReMM overview page][supremm]
in the Open XDMoD documentation.

This work was sponsored by NSF under grant numbers
[ACI 1203560][nsf-1203560], [ACI 1025159][nsf-1025159] and [ACI 1445806][nsf-1445806] for the XD Metrics Service (XMS) for NSF.

For more information, questions, feedback or bug reports send email to
`ccr-xdmod-help` at `buffalo.edu`.

Want to be notified about SUPReMM package releases and news? Subscribe to the
[XDMoD mailing list][listserv].

Software Build Requirements
---------------------------

This section provides instructions on how to create an RPM or source packages for
software development or debugging. The instructions for installing the released
packages are available on the [main website](https://supremm.xdmod.org/supremm-processing-install.html).

### Rocky Linux 8

Install the EPEL repository configuration:

    yum install epel-release

Enable the PowerTools repository (for Cython dependencies):

    sed -i 's/enabled=0/enabled=1/' /etc/yum.repos.d/Rocky-PowerTools.repo

Install the build dependencies:

    yum install -y \
        gcc \
        python3-numpy \
        python3-scipy \
        python36-devel \
        python3-Cython \
        python3-pymongo \
        python3-PyMySQL \
        python3-pytest \
        python3-pytest-cov \
        python3-mock \
        python3-pexpect \
        python3-pylint \
        python3-pcp \
        pcp-devel

Installation
------------

This project uses the [python setuptools][pydist] for package creation, and
the setup script is known to work with setuptools version 36.4.0 or later.
To install in a conda environment:

    conda create -n supremm python=3.6 cython numpy scipy
    source activate supremm
    python3 setup.py install

RPM packages are created using:

    python3 setup.py bdist_rpm


Contributing
------------

We accept contributions via standard [github pull requests][ghpr].

This project is under active development with new features planned.
Please contact us via the `ccr-xdmod-help` at `buffalo.edu` email address
before you get started so that we can co-operate and avoid duplication of effort.

Overview (for developers)
-------------------------

Full details of how to install and use the software are available on the
[SUPReMM overview page][supremm] in the Open XDMoD documentation. This section
gives a very brief overview of the summarization software for software
developers. As always, the definitive reference is the source code itself.

The summarization software processing flow is approximately as follows:

- Initial setup including parsing configuration files, opening database connections, etc.
- Query an accounting database to get the list of jobs to process
- For each job:
    - retrieve performance data that cover the time period the job ran;
    - extract the relevant datapoints per timestep;
    - run the data through the **preprocessors**;
    - run the data through the **plugins**;
    - collect the output of the **preprocessors** and **plugins** and store in an output database.

**preprocessors** and **plugins** are both python modules that implement a
defined interface. The main difference between a preprocessor and a plugin is
that the preprocessors run first and their output is available to the plugin
code.

Each **plugin** is typically responsible for generating a job-level summmary for one or many performance metrics. Each module
defines:
- an identifier for the output data;
- a list of required performance metrics;
- a mode of operation (either only process the first and last datapoints or process all data);
- an implementation of a processing function that will be called by the framework with the requested datapoints;
- an implementation of a function that will be called at the end to return the results of the analyis.

An example of a **plugin** is one that records the mean and maximum memory
usage for the job. Another example is a **plugin** that checks the temporal
variance of the L1D cache load rate to determine if the job failed prematurely.

The software that retrieves the job information from the accounting database
and writes to the output database is configurable. So, for example, you can
setup the software to write the job summary records to stdout for testing
purposes. The accounting database interface supports multiple accounting
databases (Open XDMoD being the main one).

If you are interested in doing plugin development, then a suggested starting
point is to look at some of the existing plugins. The simplest plugins, such as
the block device plugin (`supremm/plugins/Block.py`) use the framework-provided
implementation. A more complex example is the cgroup memory processor
(`supremm/plugins/CgroupMemory.py`) that contains logic to selectively
ignore certain datapoints and to do some non-trivial statistics on the data. 

If you are interested in understanding the full processing workflow, then the
starting point is the main() function in the `summarize_jobs.py` script.

License
-------

The SUPReMM processing tools package is an open source project released under
the [GNU Lesser General Public License ("LGPL") Version 3.0][lgpl3].

[lgpl3]:      http://www.gnu.org/licenses/lgpl-3.0.txt
[xdmod]:      http://xdmod.sourceforge.net/
[supremm]:    http://xdmod.sourceforge.net/supremm-overview.html
[nsf-1203560]:http://www.nsf.gov/awardsearch/showAward?AWD_ID=1203560
[nsf-1025159]:http://www.nsf.gov/awardsearch/showAward?AWD_ID=1025159
[nsf-1445806]:http://www.nsf.gov/awardsearch/showAward?AWD_ID=1445806
[listserv]:   http://listserv.buffalo.edu/cgi-bin/wa?SUBED1=ccr-xdmod-list&A=1
[ghpr]:       https://help.github.com/articles/using-pull-requests/
[pydist]:     https://setuptools.readthedocs.io/en/latest/


CHANGELOG.md
# Changelog

## [2.0.0] - 2023-09-11

### Added
- Add support for Prometheus as a performance data source. (https://github.com/ubccr/supremm/pull/268)
- Add 'datasource' field to output summarization record. (https://github.com/ubccr/supremm/pull/269)
- Add better support for the configuration check utility in `supremmconf.py`. (https://github.com/ubccr/supremm/pull/263)
- Add support for configurable job uniqueness. (https://github.com/ubccr/supremm/pull/265)

### Changed
- Update from Python 2.7 to Python 3.6.
- Update template paths to explicitly include the PCP version. (https://github.com/ubccr/supremm/pull/255)
- Use multithreaded archive indexing by default. (https://github.com/ubccr/supremm/pull/250)

### Fixed
- Update indexing in plugins to use integer division when processing jobs with greater than 64 nodes. (https://github.com/ubccr/supremm/pull/264)
- Fix string encoding from byte array to UTF-8 for PCP on RHEL 8. (https://github.com/ubccr/supremm/pull/261)

## [1.4.1] - 2020-10-14

### Fixed
- Fix edge case where one type of corrupt performance counter value was not handled
  by the hardware performance counter plugin.
- Fix bug in the job script ingestor that caused the script data in the database to 
  be set to the string "0" if a script data was ingested more than once.

## [1.4.0] - 2020-08-17

### Added
- Added new configuration setting that determines how the job script timestamps
  are interpreted. See https://supremm.xdmod.org/9.0/supremm-jobscript.html for details.
- Added timeseries plot generation for Lustre filesystem data.
- Added support for PBS style cgroups for CPU and memory metrics.

### Changed
- Improve error reporting when the temporary job directories cannot be created.
- Improve the cpu performance counter plugin to report data on the available counters
  even if not all are present.
- Update internal developer documentation
- The supremm-setup script now requires mongo client 3 or later to run the mongo database setup
  (instructions for manual setup using the older client are at https://supremm.xdmod.org/9.0/supremm-processing-configuration.html#setup-mongodb)
- Improve the job metadata handling in the developer test harness

### Fixed
- Update indexarchives.py to defer opening the connection to the XDMoD database until it
  it needed. This resolves an timeout issue when processing a large number of files
  or on a slow filesystem.

## [1.3.0] - 2019-09-30

### Added
- Added IPMI power timeseries plugin that generates timeseries data from
  IPMI power measurements.
- Add component tests for the performance counter plugins.

### Changed
- Updated RPM dependencies to match the official PCP version in RedHat/Centos 7.
- Update the CI build to test against XDMoD version 8.1.
- Improved support for hardware performance counters in the various performance
  counter plugins.

### Fixed
- Fixed bug in the performance counter preprocessor that caused it to
  incorrectly mark the performance counters as disabled by user.
- Fixed incorrect unit in schema definition for the block device timeseries metric.

## [1.2.0] - 2019-04-24

### Added
- Added GPU power plugin that summarizes the power usage for nvidia GPUs.
- Added IPMI power plugin that summarizes the power usage from impi.
- Added support for scanning the YYYY/MM/HOSTNAME/YYYY-MM-DD directory format for PCP archives.
- Added `--dry-run` option to `indexarchives.py` script (used for testing purposes).

### Changed
- Removed deprecated code that supported python 2.6.
- Updated templates to use the new recommended directory format.
- Updated test harness to work with current plugins.
- Database access code now sets the `local_infile` flag (needed for MySQL driver version 2.0.0 or later).

### Fixed
- Summarization script now only includes job start and end archives that are
  within 30 seconds of the job start and end. This mitigates a bug caused if the
  resource manager reuses the same job id for two different jobs.
- The summarization software now skips corrupt PCP archives and will try to
  continue processing the valid archives. Previously processing would stop.
- If a PCP archive has data missing at a timestep the framework will skip the timestep and
  continue processing the archive. Previously processing would stop.
- The Slurm Proc preprocessor now filters non-unicode characters from command names.
- Improvements to error handing in preprocessors and in the `perfevent` plugin.


## [1.1.0] (2) - 2018-12-07

### Fixed
- Fix dependency list for the RPM build.

## [1.1.0] - 2018-10-31

### Added
- Added support for XDMoD version 8.0.
- Added `--dry-run` option to `summarize_jobs.py` script (used for testing purposes).
- Added extra options to `summarize_jobs.py` to support more fine-grained selection of jobs to process
- Added `supremm-upgrade` script to facilitate database migrations needed for a 1.0.5 to 1.1.0 upgrade.
- Added multiprocessing support to `indexarchives.py`.
- Added option to `indexarchives.py` to estimate the archive timestamp of job level archives from the filename. This dramatically improves
 the performance on parallel filesystems that have large number of files per directory.
- Added plugin that detects periodic patterns in timeseries data.
- Added GPU usage timeseries plugin.
- Added AMD Interlagos support to the plugins that use hardware performance counters.
- Added effective CPU usage metrics to the CPU usage plugin. This generates CPU usage statistics for 
  the subset of CPUs that had any usage during a job.
- Added `summarize_mpi.py` script that uses MPI for process management. This can be used on an HPC cluster to summarize jobs in parallel across multiple compute nodes.
- Added ability to preprocess counter metrics that have < 64 bit range to 64 bit range counters.
- Added ability to call the dynamic library version of `pmlogextract`. This
  mode of operation is intended to be used when running the summarization

supremm/plugin.py
""" definition of the plugin API and implementations of some base classes that
    include common functions """

from abc import ABCMeta, abstractmethod, abstractproperty
from supremm.statistics import calculate_stats
from supremm.subsample import TimeseriesAccumulator
from supremm.errors import ProcessingError
import os
import numpy
import pkgutil

from sys import version as python_version
if python_version.startswith("2.6"):
    from backport_collections import Counter
else:
    from collections import Counter


def loadplugins(plugindir=None, namespace="plugins"):
    """ Load all of the modules from the plugins directory and instantiate the
        classes within. """

    if plugindir == None:
        plugindir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "plugins")

    analytics = []

    for _, x, _ in pkgutil.iter_modules([plugindir]):
        m1 = __import__('supremm.' + namespace + '.' + x)
        m = getattr(m1, namespace)
        m = getattr(m, x)
        c = getattr(m, x)
        if issubclass(c, Plugin) or issubclass(c, PreProcessor):
            analytics.append(c)
        else:
            del m1

    return analytics


def loadpreprocessors(preprocdir=None):
    """ preprocessors have same api as plugins. The difference is that they
        are called first so the results are available to the plugins.
        There is no dependency management as of yet so a preproc cannot require
        another preproc.
    """
    if preprocdir == None:
        preprocdir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "preprocessors")

    return loadplugins(preprocdir, "preprocessors")

class NodeMetadata(object):
    """ Wrapper class that contains info about a job node. This is passed to
        the process function of the plugin. """
    __metaclass__ = ABCMeta

    @property
    @abstractmethod
    def nodename(self):
        """ Returns a unique string identifier for the node (typically the hostname) """
        pass

    @property
    @abstractmethod
    def nodeindex(self):
        """ returns a unique numerical identifier for the node """
        pass

class Plugin(object):
    """ abstract base class describing the plugin interface """
    __metaclass__ = ABCMeta

    def __init__(self, job):
        self._job = job
        self._status = "uninitialized"

    @property
    def status(self):
        """ The status state is used by the framework to decide whether to include the
            plugin data in the output """
        return self._status

    @status.setter
    def status(self, value):
        """ status can be set by the framework """
        self._status = value

    @abstractmethod
    def process(self, nodemeta, timestamp, data, description):
        """ process is called for every requested data point """
        pass

    @abstractmethod
    def results(self):
        """ results will be called once after all the datapoints have had calls to  process()"""
        pass

    @abstractproperty
    def name(self):
        pass

    @abstractproperty
    def requiredMetrics(self):
        pass

    @abstractproperty
    def optionalMetrics(self):
        pass

    @abstractproperty
    def derivedMetrics(self):
        pass

    @abstractproperty
    def mode(self):
        pass


class PreProcessor(object):
    """
    Preprocessors are called on each archive before all of the plugins are
    run. The results of the preprocessor are therefore available to the plugins.
    The preprocessor results should be added to the job object by the preprocessor
    using the job.addata() function.
    """
    __metaclass__ = ABCMeta

    def __init__(self, job):
        self._job = job
        self._status = "uninitialized"

    @property
    def status(self):
        """ The status state is used by the framework to decide whether to include the
            plugin data in the output """
        return self._status

    @status.setter
    def status(self, value):
        """ status can be set by the framework """
        self._status = value

    @abstractmethod
    def hoststart(self, hostname):
        """ Called by the framework for all hosts assigned to the job whether or not they contain
            the required metrics.
        """
        pass

    @abstractmethod
    def process(self, timestamp, data, description):
        """ Called by the framework for all datapoints collected for the host
            referenced in the preceeding call to hoststart. It a host has no data
            then this will not be called.
        """
        pass

    @abstractmethod
    def results(self):
        """ preprocessors may return a results object that will be added to the summary
        document.
        """
        pass

    @abstractmethod
    def hostend(self):
        """ Called after all of the data available for a host has been processed. """
        pass

    @abstractproperty
    def name(self):
        pass

    @abstractproperty
    def requiredMetrics(self):
        pass

    @abstractproperty
    def optionalMetrics(self):
        pass

    @abstractproperty
    def derivedMetrics(self):
        pass

    @abstractproperty
    def mode(self):
        pass


class DeviceBasedPlugin(Plugin):
    """
    A base abstract class for summarising the job-delta for device-based metrics
    The plugin name and list of required metrics must be provided by the implementation
    """
    __metaclass__ = ABCMeta

    mode = property(lambda x: "firstlast")

    def __init__(self, job):
        super(DeviceBasedPlugin, self).__init__(job)
        self._first = {}
        self._data = {}
        self._error = None
        self.allmetrics = self.requiredMetrics + self.optionalMetrics

    def process(self, nodemeta, timestamp, data, description):

        if len(data[0]) == 0:
            return False

        if nodemeta.nodename not in self._first:
            self._first[nodemeta.nodename] = numpy.array(data)
            return True

        ndata = numpy.array(data)

        if ndata.shape != self._first[nodemeta.nodename].shape:
            self._error = ProcessingError.INDOMS_CHANGED_DURING_JOB
            return False

        hostdata = ndata - self._first[nodemeta.nodename]

        for mindex, i in enumerate(description):
            for index in xrange(len(hostdata[mindex, :])):
                indom = i[1][index]
                metricname = self.allmetrics[mindex]

                if indom not in self._data:
                    self._data[indom] = {}
                if metricname not in self._data[indom]:
                    self._data[indom][metricname] = []
                self._data[indom][metricname].append(hostdata[mindex, index])

        return True

    def results(self):

        if self._error != None:
            return {"error": self._error}

        if len(self._data) == 0:
            return {"error": ProcessingError.INSUFFICIENT_DATA}

        output = {}

        for devicename, device in self._data.iteritems():
            cleandevname = devicename.replace(".", "-")
            output[cleandevname] = {}
            for metricname, metric in device.iteritems():
                prettyname = "-".join(metricname.split(".")[2:])
                output[cleandevname][prettyname] = calculate_stats(metric)

        return output

class DeviceInstanceBasedPlugin(Plugin):
    """
    A base abstract class for summarising the job-delta for device-based metrics
    that only have a singe instance.
    The plugin name and list of required metrics must be provided by the implementation
    """
    __metaclass__ = ABCMeta

    mode = property(lambda x: "firstlast")

    def __init__(self, job):
        super(DeviceInstanceBasedPlugin, self).__init__(job)
        self._first = {}
        self._data = {}
        self._error = None

    def process(self, nodemeta, timestamp, data, description):

        if len(data[0]) == 0:
            return False

        if nodemeta.nodeindex not in self._first:
            self._first[nodemeta.nodeindex] = numpy.array(data)
            return True

        hostdata = numpy.array(data) - self._first[nodemeta.nodeindex]

        for idx, metricname in enumerate(self.requiredMetrics):
            if metricname not in self._data:
                self._data[metricname] = []
            self._data[metricname].append(hostdata[idx, 0])

    def results(self):

        if self._error != None:
            return {"error": self._error}

        if len(self._data) == 0:
            return {"error": ProcessingError.INSUFFICIENT_DATA}

        output = {}

        for metricname, metric in self._data.iteritems():
            prettyname = "-".join(metricname.split(".")[1:])
            output[prettyname] = calculate_stats(metric)

        return output


class RateConvertingTimeseriesPlugin(Plugin):
    """ 
    A base abstract class for generating a timeseries summary for values that should
    be converted to rates, one per node.
    The plugin name,  list of required metrics and generator function must be provided by the implementation
    """
    __metaclass__ = ABCMeta

    mode = property(lambda x: "timeseries")

    def __init__(self, job):
        super(RateConvertingTimeseriesPlugin, self).__init__(job)
        self._data = TimeseriesAccumulator(job.nodecount, self._job.walltime)
        self._hostdata = {}

    @abstractmethod
    def computetimepoint(self, data):
        """ Called with the data for each timepoint on each host """
        pass

    def process(self, nodemeta, timestamp, data, description):

        if nodemeta.nodeindex not in self._hostdata:
            self._hostdata[nodemeta.nodeindex] = 1

        datum = self.computetimepoint(data)
        if datum != None:
            self._data.adddata(nodemeta.nodeindex, timestamp, datum)

    def results(self):

        if len(self._hostdata) != self._job.nodecount:
            return {"error": ProcessingError.INSUFFICIENT_HOSTDATA}

        values = self._data.get()

        if len(values[0, :, 0]) < 3:
            return {"error": ProcessingError.JOB_TOO_SHORT}

        rates = numpy.diff(values[:, :, 1]) / numpy.diff(values[:, :, 0])

        if len(self._hostdata) > 64:

            # Compute min, max & median data and only save the host data
            # for these hosts

            sortarr = numpy.argsort(rates.T, axis=1)

            retdata = {
                "min": self.collatedata(sortarr[:, 0], rates),
                "max": self.collatedata(sortarr[:, -1], rates),
                "med": self.collatedata(sortarr[:, sortarr.shape[1] / 2], rates),
                "times": values[0, 1:, 0].tolist(),
                "hosts": {}
            }

            uniqhosts = Counter(sortarr[:, 0])
            uniqhosts.update(sortarr[:, -1])
            uniqhosts.update(sortarr[:, sortarr.shape[1] / 2])
            includelist = uniqhosts.keys()
        else:
            # Save data for all hosts
            retdata = {
                "times": values[0, 1:, 0].tolist(),
                "hosts": {}
            }
            includelist = self._hostdata.keys()


        for hostidx in includelist:
            retdata['hosts'][str(hostidx)] = {}
            retdata['hosts'][str(hostidx)]['all'] = rates[hostidx, :].tolist()

        return retdata

    @staticmethod
    def collatedata(args, rates):
        """ build output data """
        result = []
        for timepoint, hostidx in enumerate(args):
            try:
                result.append([rates[hostidx, timepoint], int(hostidx)])
            except IndexError:
                pass

        return result
